File: src/tmp/Python/tree/binary_search_tree_iterator.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BSTIterator:

    def __init__(self, root: TreeNode):
        self.stack = list()
        while root:
            self.stack.append(root)
            root = root.left


    def next(self) -> int:
        point = self.stack.pop()
        temp = point.right
        while temp:
            self.stack.append(temp)
            temp = temp.left
        
        return point.val

    def hasNext(self) -> bool:
        return len(self.stack) != 0

File: core/tmp/Java/src/com/dbc/algorithm/Tree/BinarySearchTreeIterator.java
package com.dbc.algorithm.Tree;

import java.util.Stack;

public class BinarySearchTreeIterator {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private Stack<TreeNode> stack;
    public BinarySearchTreeIterator(TreeNode root) {
        stack = new Stack<>();
        while (root != null) {
            stack.push(root);
            root = root.left;
        }
    }

    public int next() {
        TreeNode point = stack.pop();
        TreeNode temp = point.right;
        while (temp != null) {
            stack.push(temp);
            temp = temp.left;
        }
        return point.val;
    }

    public boolean hasNext() {
        return !stack.isEmpty();
    }
}
